### ISSUE TITLE
Cumulativity not working for data/record types

### ISSUE THREAD

**Issue by rw111** [agda/agda#6726]
Using latest official version of Agda (and also in Cubical Agda) cumulativity fails to work in some situations. 

Here is some code to illustrate:

```agda
{-# OPTIONS --guardedness #-}
{-# OPTIONS --no-main #-}
{-# OPTIONS --cumulativity #-}

module agdaCumulativity where

open import Agda.Primitive

record Tlist {ℓ : Level} : Set (lsuc ℓ)  where
  constructor _,_
  coinductive
  field
    hd : Set ℓ
    tl : Tlist {ℓ}

open Tlist

egTlist : Tlist {lzero}
hd (egTlist) = Bool
tl (egTlist) = egTlist

egTlist3 : Tlist {lsuc (lsuc (lsuc lzero))}
hd egTlist3 = Set₁                -- this is lifted correctly
tl egTlist3 = egTlist              -- THE IMPLICIT LEVEL lzero DOESN'T GET LIFTED TO THE REQUIRED LEVEL  {lsuc (lsuc (lsuc lzero))}
```

Specifically here `egTlist` is defined as having an implicit level argument of `lzero`, but this fails to be lifted to the required level of `lsuc (lsuc (lsuc lzero))` in the definition of `egTlist3` in the `tl` line. The `hd` line is working fine, and you can put `Bool` or `Set 2` in place of the `Set 1` here, and all cases will be assimilated correctly to level  `lsuc (lsuc (lsuc lzero))`.

**Comment by jespercockx** [comment]
This is just the current (poor) heuristic for cumulativity kicking in, see https://github.com/agda/agda/issues/4089

**Comment by arthur-adjedj** [comment]
I don't believe this issue has to do with the current heuristics referred to in #4089 . No level metavariables are generated here, and everything seems to be explicit, so it shouldn't be an artifact from the janky unification. The error that this code gives is that `lzero != lsuc (lsuc (lsuc lzero))`, which makes me believe the conversion might simply have kicked off with the wrong Comparison tag here when comparing the type of `egTlist` to `egTlist3`'s.

**Comment by rw111** [comment]
If so Arthur could this part of the solver "generalised _explicit_ lifting" be made efficient and safe? I mean, I sort of suspect that the real problems with the solver are coming from the fact that without explicitly knowing or stating the target level we have to deal with possibly ambiguous solutions as in the #Cubical "Cumulativity" thread.

**Comment by andreasabel** [comment]
@rw111: For your example to work, don't you need cumulative (co)inductive types as they have in Coq: https://drops.dagstuhl.de/opus/volltexte/2018/9199/pdf/LIPIcs-FSCD-2018-29.pdf
Agda does not have these yet, AFAIK.

**Comment by rw111** [comment]
Why? You can do it with extra boiler plate code, so the answer to your question is therefore 'no'. Coq is designed differently. It may or may not apply. Given that this is such a basic thing to do with cumulativity (it was literally my first usage of it), it's still a bug.

**Comment by andreasabel** [comment]
@rw1111 wrote:
> You can do it with extra boiler plate code,

Could you spell out the extra code needed to pass the MWE in the OP?
Currently, the OP does not spell out the additions that would make the example check in Agda.

**Comment by rw111** [comment]
Sure. My non-cumulativity rewrite is in an Agda Zulip thread (entry July 13):  #general "is this a bug in cumulativity?"

Here it is for the record:

```agda
{-# OPTIONS --safe #-}
{-# OPTIONS --guardedness #-}
{-# OPTIONS --no-main #-}

--{-# OPTIONS --cumulativity #-}

module agdaCumulativity where

open import Data.Bool
open import Agda.Primitive
open import Level

record Tlist {ℓ : Level} : Set (lsuc ℓ)  where
  constructor _,_
  coinductive
  field
    hd : Set ℓ
    tl : Tlist {ℓ}

open Tlist

egTlist : Tlist {lzero}
hd (egTlist) = Bool
tl (egTlist) = egTlist

HeavyLift : {ℓ : Level} -> (ℓ' : Level) -> Tlist {ℓ} → Tlist {ℓ ⊔ ℓ'}
hd (HeavyLift {ℓ} ℓ' x) = Lift ℓ' (hd x)
tl (HeavyLift {ℓ} ℓ' x) = HeavyLift ℓ' (tl x)

egTlist3 : Tlist {lsuc (lsuc (lsuc lzero))}
hd egTlist3 = Lift (suc (suc (suc zero))) Set₁
tl egTlist3 = HeavyLift (suc (suc (suc zero))) egTlist                  -- THIS LAST LINE NEEDS A SPECIAL HELPER FUNCTION
```

More importantly it follows from Adjedj's comment immediately before on July 11th that it _is_ a bug (or at least 'almost certainly'...). He also says he is looking at it. 

Thing is, 'cumulativity' isn't really cumulativity if it can't do this. So the other side of this is that if you are right, and you may be, then cumulativity needs a rethink.

**Comment by arthur-adjedj** [comment]
I'd like to point out that I'm no authority on the matter, and my comment is a personal opinion, though I am planning on trying to fix this. I may be missing something, but to me, the code given in the first comment of the issue [here](https://github.com/agda/agda/issues/6726#issue-1803029344) looks like something that should be accepted by agda, knowing that the "inductive equivalent" is accepted with no issue:
```agda
{-# OPTIONS --guardedness #-}
{-# OPTIONS --no-main #-}
{-# OPTIONS --cumulativity #-}

open import Data.Bool
open import Agda.Primitive

data Tlist {ℓ : Level} : Set (lsuc ℓ)  where
    nil  : Tlist
    cons : Set ℓ -> Tlist -> Tlist

open Tlist

egTlist : Tlist {lzero}
egTlist = cons Bool nil

egTlist3 : Tlist {lsuc (lsuc (lsuc lzero))}
egTlist3 = cons Set₁ egTlist -- no error
```

**Comment by andreasabel** [comment]
@rw111: Thanks for the update.  
The fact that you need a recursive function to make the OP work actually confirms my hunch that this is rather a new feature (strongly reminds me on cumulativity in CIC---citation above) than just a bug fix.

Contrastingly, the example given by @arthur-adjedj does not need any recursion.  
There, it is plainly the fact that with `--cumulativity`, `Set` is monotone in the level argument.  That's all this feature promises at the moment.
